src/BaGet.Core/Indexing/IPackageIndexingMiddleware.cs

```diff
@@ -0,0 +1,73 @@
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace BaGet.Core
+{
+    /// <summary>
+    /// A single indexing step for <see cref="IPackageIndexingService.IndexAsync(Stream, CancellationToken)"/>.
+    /// </summary>
+    public interface IPackageIndexingMiddleware
+    {
+        /// <summary>
+        /// Attempt to index a package.
+        /// </summary>
+        /// <param name="context">The context for the package that is being indexed.</param>
+        /// <param name="next">The next middleware in the indexing chain.</param>
+        /// <returns>The result of indexing the package.</returns>
+        Task<PackageIndexingResult> IndexAsync(PackageIndexingContext context, PackageIndexingDelegate next);
+    }
+
+    /// <summary>
+    /// A delegate that wraps a <see cref="IPackageIndexingMiddleware"/>.
+    /// </summary>
+    /// <returns>A task that completes when the middleware is done indexing the package.</returns>
+    public delegate Task<PackageIndexingResult> PackageIndexingDelegate();
+
+    /// <summary>
+    /// The state of the package indexing operation.
+    /// </summary>
+    public class PackageIndexingContext : IDisposable
```

Comment:
Move this to its own class. Consider making it abstract with a `DefaultPackagingIndexingContext` (similar to HttpContext and `DefaultHttpContext`)